File: EliteFramework/UIImage/EliteImage.m
//
//  EliteCategories.m
//  BLEDrawingPad
//
//  Created by Lifelong-Study on 2015/4/21.
//  Copyright (c) 2015年 PaoYo Ding. All rights reserved.
//

#import "EliteImage.h"

@implementation UIImage (EliteImageCategory)

+ (UIImage *)imageWithView:(UIView *)view
{
    UIGraphicsBeginImageContextWithOptions(view.bounds.size, NO, 1.0);
    
    [view.layer renderInContext:UIGraphicsGetCurrentContext()];
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
    
    return image;
}

+ (UIImage *)imageWithColor:(UIColor *)color
{
    CGRect rect = CGRectMake(0.0f, 0.0f, 1.0f, 1.0f);
    
    UIGraphicsBeginImageContext(rect.size);
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    CGContextSetFillColorWithColor(context, [color CGColor]);
    
    CGContextFillRect(context, rect);
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    
    CGContextRelease(context);
    
    UIGraphicsEndImageContext();
    
    return image;
}

+ (UIImage *)imageWithLayer:(CALayer *)layer
{
    UIGraphicsBeginImageContext([layer frame].size);
    
    [layer renderInContext:UIGraphicsGetCurrentContext()];
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    
    UIGraphicsEndImageContext();
    
    return image;
}

//
- (UIImage *)scaledToSize:(CGSize)size
{
    // Create a graphics image context
    UIGraphicsBeginImageContext(size);
    
    // Tell the old image to draw in this new context, with the desired
    // new size
    [[self copy] drawInRect:CGRectMake(0, 0, size.width, size.height)];
    
    // Get the new image from the context
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    
    // End the context
    UIGraphicsEndImageContext();
    
    // Return the new image.
    return image;
}

// 影像壓縮
// maxWidth : 影像的最大寬度
// maxHeight: 影像的最大高度
// quality  : 影像品質的壓縮比( 0.0 ~ 1.0 )
- (UIImage *)imageCompressWithMaxWidth:(CGFloat)maxWidth maxHeight:(CGFloat)maxHeight quality:(CGFloat)quality
{
    float actualWidth  = self.size.width;
    float actualHeight = self.size.height;
    float maxRatio     = maxWidth / maxHeight;
    float imgRatio     = actualWidth/actualHeight;
    //    float compressionQuality = 0.5; // 影像品質的壓縮比 0.0 ~ 1.0
    
    if (actualHeight > maxHeight || actualWidth > maxWidth) {
        if (NO) {
        } else if (imgRatio < maxRatio) {
            // 根據  maxHeight 調節影像寬度
            imgRatio = maxHeight / actualHeight;
            actualWidth = imgRatio * actualWidth;
            actualHeight = maxHeight;
        } else if(imgRatio > maxRatio) {
            // 根據  maxWidth 調節影像高度
            imgRatio = maxWidth / actualWidth;
            actualHeight = imgRatio * actualHeight;
            actualWidth = maxWidth;
        } else {
            actualHeight = maxHeight;
            actualWidth = maxWidth;
        }
    }
    
    CGRect rect = CGRectMake(0.0, 0.0, actualWidth, actualHeight);
    UIGraphicsBeginImageContext(rect.size);
    [self drawInRect:rect];
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    NSData *imageData = UIImageJPEGRepresentation(image, quality);
    UIGraphicsEndImageContext();
    
    return [UIImage imageWithData:imageData];
}

- (UIImage *)clearColor:(UIColor *)color
{
    return [self changeColor:color toColor:[UIColor clearColor]];
}

- (UIImage *)changeColor:(UIColor *)fromColor toColor:(UIColor *)toColor
{
    CGImageRef originalImage   = self.CGImage;
    
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    
    CGContextRef bitmapContext = CGBitmapContextCreate(NULL,
                                                       CGImageGetWidth(originalImage),
                                                       CGImageGetHeight(originalImage), 8,
                                                       CGImageGetWidth(originalImage) * 4,
                                                       colorSpace,
                                                       kCGImageAlphaPremultipliedLast);
    CGColorSpaceRelease(colorSpace);
    
    CGContextDrawImage(bitmapContext, CGRectMake(0, 0, CGBitmapContextGetWidth(bitmapContext),CGBitmapContextGetHeight(bitmapContext)), originalImage);
    
    UInt8 *data = CGBitmapContextGetData(bitmapContext);
    
    
    unsigned long bytes = CGBitmapContextGetHeight(bitmapContext) * CGBitmapContextGetBytesPerRow(bitmapContext);
    
    
    double inRed   = 0.0, inGreen   = 0.0, inBlue   = 0.0, inAlpha   = 0.0;
    double fromRed = 0.0, fromGreen = 0.0, fromBlue = 0.0, fromAlpha = 0.0;
    double toRed   = 0.0, toGreen   = 0.0, toBlue   = 0.0, toAlpha   = 0.0;
    
    // Get RGB values of fromColor
    unsigned long fromCountComponents = CGColorGetNumberOfComponents([fromColor CGColor]);
    if (fromCountComponents == 4) {
        const CGFloat *_components = CGColorGetComponents([fromColor CGColor]);
        fromRed   = _components[0];
        fromGreen = _components[1];
        fromBlue  = _components[2];
        fromAlpha = _components[3];
    }
    
    
    // Get RGB values for toColor
    unsigned long toCountComponents = CGColorGetNumberOfComponents([toColor CGColor]);
    if (toCountComponents == 4) {
        const CGFloat *_components = CGColorGetComponents([toColor CGColor]);
        toRed   = _components[0];
        toGreen = _components[1];
        toBlue  = _components[2];
        toAlpha = _components[3];
    }
    
    //
    for (NSInteger i = 0 ; i < bytes; i += 4) {
        
        //
        inRed   = data[i + 0] / 255.0;
        inGreen = data[i + 1] / 255.0;
        inBlue  = data[i + 2] / 255.0;
        inAlpha = data[i + 3] / 255.0;
        
        //
        if (inRed == fromRed && inGreen == fromGreen && inBlue == fromBlue && inAlpha == fromAlpha) {
            data[i + 0] = toRed;
            data[i + 1] = toGreen;
            data[i + 2] = toBlue;
            data[i + 3] = toAlpha;
        }
    }
    
    CGImageRef outImage = CGBitmapContextCreateImage(bitmapContext);
    
    UIImage *image = [UIImage imageWithCGImage:outImage];
    
    CGImageRelease(outImage);
    
    return image;
}

static inline double radians (double degrees) {return degrees * M_PI/180;}

- (UIImage *)rotate:(UIImageOrientation)orientation
{
    UIGraphicsBeginImageContext(self.size);
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    
    if (orientation == UIImageOrientationRight) {
        CGContextRotateCTM (context, radians(90));
    } else if (orientation == UIImageOrientationLeft) {
        CGContextRotateCTM (context, radians(-90));
    } else if (orientation == UIImageOrientationDown) {
        // NOTHING
    } else if (orientation == UIImageOrientationUp) {
        CGContextRotateCTM (context, radians(90));
    }
    
    [self drawAtPoint:CGPointMake(0, 0)];
    
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image;
}

// 影像剪裁
- (UIImage *)trimImageWithMask:(CGRect)maskFrame
{
    CGImageRef imageRef = CGImageCreateWithImageInRect(self.CGImage, maskFrame);
    UIImage *croppedImage = [UIImage imageWithCGImage:imageRef];
    CGImageRelease(imageRef);
    
    return croppedImage;
}

@end

File: EliteFramework/EliteImage.swift
//
//  EliteImage.swift
//  EliteFramework
//
//  Created by Lifelong-Study on 2016/3/9.
//  Copyright © 2016年 Lifelong-Study. All rights reserved.
//

import UIKit


struct RGBA {
    private var color: UInt32
    
    var redComponent: UInt8 {
        return UInt8((color >> 24) & 255)
    }
    
    var greenComponent: UInt8 {
        return UInt8((color >> 16) & 255)
    }
    
    var blueComponent: UInt8 {
        return UInt8((color >> 8) & 255)
    }
    
    var alphaComponent: UInt8 {
        return UInt8((color >> 0) & 255)
    }
    
    init(red: UInt8, green: UInt8, blue: UInt8, alpha: UInt8) {
        let red   = UInt32(red)
        let green = UInt32(green)
        let blue  = UInt32(blue)
        let alpha = UInt32(alpha)
        color = (red << 24) | (green << 16) | (blue << 8) | (alpha << 0)
    }
    
    static let red     = RGBA(red: 255, green: 0,   blue: 0,   alpha: 255)
    static let green   = RGBA(red: 0,   green: 255, blue: 0,   alpha: 255)
    static let blue    = RGBA(red: 0,   green: 0,   blue: 255, alpha: 255)
    static let white   = RGBA(red: 255, green: 255, blue: 255, alpha: 255)
    static let black   = RGBA(red: 0,   green: 0,   blue: 0,   alpha: 255)
    static let magenta = RGBA(red: 255, green: 0,   blue: 255, alpha: 255)
    static let yellow  = RGBA(red: 255, green: 255, blue: 0,   alpha: 255)
    static let cyan    = RGBA(red: 0,   green: 255, blue: 255, alpha: 255)
    static let clear   = RGBA(red: 0,   green: 0,   blue: 0,   alpha: 0)
    
    static let bitmapInfo = CGImageAlphaInfo.premultipliedLast.rawValue | CGBitmapInfo.byteOrder32Little.rawValue
    
    static func ==(lhs: RGBA, rhs: RGBA) -> Bool {
        return lhs.color == rhs.color
    }
}

public extension UIImage {
    
    //
    fileprivate convenience init(image: UIImage?) {
        if let cgimage = image?.cgImage {
            self.init(cgImage: cgimage)
        } else {
            self.init()
        }
    }
    
    // initialization
    convenience init(color: UIColor) {
        
        let rect = CGRect(x: 0.0, y: 0.0, width: 1.0, height: 1.0)
        
        UIGraphicsBeginImageContextWithOptions(rect.size, false, 0.0)
        
        color.setFill()
        
        UIRectFill(rect)
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        
        UIGraphicsEndImageContext()
        
        self.init(image: image)
    }
    
    convenience init(layer: CALayer) {
        UIGraphicsBeginImageContextWithOptions(layer.bounds.size, false, 0.0)
        
        if let context = UIGraphicsGetCurrentContext() {
            layer.render(in: context)
        }
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        
        UIGraphicsEndImageContext()
        
        self.init(image: image)
    }
    
    
    convenience init(render view: UIView, from: RenderDirection, to: RenderDirection, colors: [UIColor]) {
        let layer = CALayer()
        
        layer.bounds = view.bounds
        
        layer.renderGradient(from: from, to: to, colors: colors)
        
        self.init(layer: layer)
    }
    
    // function
    func radians(_ degrees: Double) -> CGFloat {
        return CGFloat(degrees * Double.pi / 180.0)
    }
    
    /*! 影像剪裁 */
    func trimImageWithMask(_ maskFrame: CGRect) -> UIImage? {
        if let cgimage = self.cgImage {
            if let imageRef = cgimage.cropping(to: maskFrame) {
                return UIImage(cgImage: imageRef)
            }
        }
        
        return nil
    }
    
    //!
    func scaleToSize(size: CGSize) -> UIImage {
        UIGraphicsBeginImageContext(size);
        
        draw(in: CGRect(x: 0.0, y: 0.0, width: size.width, height: size.height))
        
        let image = UIGraphicsGetImageFromCurrentImageContext();
        
        UIGraphicsEndImageContext();
        
        return image!;
    }
    
    //!
    func compress(maxWidth: Double, maxHeight: Double, quality: Double) -> UIImage? {
        
        var actualWidth  = Double(self.size.width)
        var actualHeight = Double(self.size.height)
        let maxRatio     = maxWidth / maxHeight;
        var imgRatio     = actualWidth / actualHeight;
        
        
        if actualHeight > maxHeight || actualWidth > maxWidth {
            if false {
            } else if imgRatio < maxRatio {
                // 根據  maxHeight 調節影像寬度
                imgRatio = maxHeight / actualHeight;
                actualWidth = imgRatio * actualWidth;
                actualHeight = maxHeight;
            } else if imgRatio > maxRatio {
                // 根據  maxWidth 調節影像高度
                imgRatio = maxWidth / actualWidth;
                actualHeight = imgRatio * actualHeight;
                actualWidth = maxWidth;
            } else {
                actualHeight = maxHeight;
                actualWidth = maxWidth;
            }
        }
        
        
        let rect = CGRect(x: 0.0, y: 0.0, width: CGFloat(actualWidth), height: CGFloat(actualHeight))
        UIGraphicsBeginImageContext(rect.size)
        draw(in: rect)
        let image = UIGraphicsGetImageFromCurrentImageContext()
        let imageData = image?.jpegData(compressionQuality: CGFloat(quality))
        UIGraphicsEndImageContext()
        
        
        return UIImage(data: imageData!)
    }
    
    //!
    func rotate(orientation: UIImage.Orientation) -> UIImage? {
        
        if orientation != .right && orientation != .left {
            return self
        }
        
        UIGraphicsBeginImageContext(self.size)
        
        let context = UIGraphicsGetCurrentContext()
        
        if false {
        } else if orientation == .right {
            context!.rotate(by: radians(90))
        } else if orientation == .left {
            context!.rotate(by: radians(-90))
        }
        
        draw(at: CGPoint(x: 0, y: 0))
        
        let image = UIGraphicsGetImageFromCurrentImageContext()
        
        UIGraphicsEndImageContext()
        
        return image
    }
    
    // 將圖片上的 from 的替換成 to 的顏色
    func replaceColor(from: UIColor, to: UIColor) -> UIImage? {
        var fromR: CGFloat = 0.0, fromG: CGFloat = 0.0, fromB: CGFloat = 0.0, fromA: CGFloat = 0.0
        var   toR: CGFloat = 0.0,   toG: CGFloat = 0.0,   toB: CGFloat = 0.0,   toA: CGFloat = 0.0
        
        from.getRed(&fromR, green: &fromG, blue: &fromB, alpha: &fromA)
        to.getRed(&toR, green: &toG, blue: &toB, alpha: &toA)
        
        let fromRGBA = RGBA.init(red: UInt8(fromR * 255.0), green: UInt8(fromB * 255.0), blue: UInt8(fromB * 255.0), alpha: UInt8(fromA * 255.0))
        let   toRGBA = RGBA.init(red: UInt8(toR * 255.0),   green: UInt8(toG * 255.0),   blue: UInt8(toB * 255.0),   alpha: UInt8(toA * 255.0))
        
        return replaceColor(from: fromRGBA, to: toRGBA)
    }
    
    private func replaceColor(from: RGBA, to: RGBA) -> UIImage? {
        guard let inputCGImage = self.cgImage else {
            print("unable to get cgImage")
            return nil
        }
        
        let colorSpace       = CGColorSpaceCreateDeviceRGB()
        let width            = inputCGImage.width
        let height           = inputCGImage.height
        let bytesPerPixel    = 4
        let bitsPerComponent = 8
        let bytesPerRow      = bytesPerPixel * width
        let bitmapInfo       = RGBA.bitmapInfo
        
        guard let context = CGContext(data: nil, width: width, height: height, bitsPerComponent: bitsPerComponent, bytesPerRow: bytesPerRow, space: colorSpace, bitmapInfo: bitmapInfo) else {
            print("unable to create context")
            return nil
        }
        context.draw(inputCGImage, in: CGRect(x: 0, y: 0, width: width, height: height))
        
        guard let buffer = context.data else {
            print("unable to get context data")
            return nil
        }
        
        let pixelBuffer = buffer.bindMemory(to: RGBA.self, capacity: width * height)
        
        for row in 0 ..< Int(height) {
            for column in 0 ..< Int(width) {
                let offset = row * width + column
                if pixelBuffer[offset] == from {
                    pixelBuffer[offset] = to
                }
            }
        }
        
        if let outputCGImage = context.makeImage() {
            return UIImage(cgImage: outputCGImage, scale: scale, orientation: imageOrientation)
        } else {
            return nil
        }
    }
    
}

